src/PsrLogAdapter.php

```diff
@@ -0,0 +1,161 @@
+        /* libmongoc considers "critical" less severe than "error" so map it to
+         * "error" in the PSR logger. */
+        LogSubscriber::LEVEL_CRITICAL => LogLevel::ERROR,
+        LogSubscriber::LEVEL_WARNING => LogLevel::WARNING,
+        LogSubscriber::LEVEL_MESSAGE => LogLevel::NOTICE,
+        LogSubscriber::LEVEL_INFO => LogLevel::INFO,
+        LogSubscriber::LEVEL_DEBUG => LogLevel::DEBUG,
+    ];
+
+    public static function addLogger(LoggerInterface $logger): void
+    {
+        $instance = self::getInstance();
+
+        $instance->loggers->attach($logger);
+
+        addSubscriber($instance);
+    }
+
+    /**
+     * Forwards a log message from libmongoc/PHPC to all registered PSR loggers.
+     *
+     * @see LogSubscriber::log()
+     */
+    public function log(int $mongocLevel, string $domain, string $message): void
+    {
+        if (! isset(self::MONGOC_TO_PSR[$mongocLevel])) {
+            throw new UnexpectedValueException(sprintf(
+                'Expected level to be >= %d and <= %d, %d given for domain "%s" and message: %s',
+                LogSubscriber::LEVEL_ERROR,
+                LogSubscriber::LEVEL_DEBUG,
+                $mongocLevel,
+                $domain,
+                $message,
+            ));
+        }
+
+        $instance = self::getInstance();
+        $psrLevel = self::MONGOC_TO_PSR[$mongocLevel];
+        $context = ['domain' => $domain];
+
+        foreach ($instance->loggers as $logger) {
+            $logger->log($psrLevel, $message, $context);
+        }
+    }
+
+    public static function removeLogger(LoggerInterface $logger): void
+    {
+        $instance = self::getInstance();
+        $instance->loggers->detach($logger);
+
+        if ($instance->loggers->count() === 0) {
+            removeSubscriber($instance);
+        }
+    }
+
+    /**
+     * Writes a log message to all registered PSR loggers.
+     *
+     * This function is intended for internal use within the library.
+     */
+    public static function writeLog(int $specLevel, string $domain, string $message): void
+    {
+        if (! isset(self::SPEC_TO_PSR[$specLevel])) {
+            throw new UnexpectedValueException(sprintf(
+                'Expected level to be >= %d and <= %d, %d given for domain "%s" and message: %s',
+                self::EMERGENCY,
+                self::TRACE,
+                $specLevel,
+                $domain,
+                $message,
+            ));
+        }
+
+        $instance = self::getInstance();
+        $psrLevel = self::SPEC_TO_PSR[$specLevel];
+        $context = ['domain' => $domain];
+
+        foreach ($instance->loggers as $logger) {
+            $logger->log($psrLevel, $message, $context);
+        }
+    }
+
+    private function __construct()
+    {
+        $this->loggers = new SplObjectStorage();
+    }
+
+    private static function getInstance(): self
+    {
+        return self::$instance ??= new self();
```

Comment:
@GromNaN: Also refactored this per your previous suggestion in https://github.com/mongodb/mongo-php-library/pull/1158#discussion_r1324991480.